### ISSUE TITLE
Attachment insert never more works

### ISSUE THREAD

**Issue by PortaMx** [SimpleMachines/SMF#3381]
I download the archive today and see, that the image attachment not more works.

If I attach a Image and then insert this in the message, I see the message
**No message is associated with this attachment.**
and the image is only shown as attach.

On Preview I get the error
**Undefined index: name=image11.png type=image/png]5
File: /Sources/Subs-Attachments.php
Line: 904**

**Comment by Oldiesmann** [comment]
The "no message is associated with this attachment" thing when inserting just an attachment is a known issue. I'll look into the preview error.

**Comment by Oldiesmann** [comment]
It looks like the error on preview is due to a problem with how the attach bbcode tag is parsed.

This is the code on line 904:
`$attachContext = $allAttachments[0][$attachID];`

$attachID is a parameter sent to that function, which is called from line 1101 of Subs.php:
`$currentAttachment = parseAttachBBC($attachID);`

$attachID is set from the $data parameter passed to the tag's validate function:
`$attachID = $data;`

**Comment by PortaMx** [comment]
If this bug fixed now?
Currently I can't insert a image with the attach bbc tag.
The image is attached, but not inserted in the message, only the text string is shown ..
Why you changed this? In the  last testversion it's worked perfect ...

**Comment by MissAllSunday** [comment]
Will take a look at it, the problem with this inline atachment stuff is if one thing gets fixed another gets borked...   the 0 key in $allAttachments[0]  is used to preview attachments that hasn't been publish yet (ie no msg id) thats the only sane way I found to be able to preview new attachments.

**Comment by PortaMx** [comment]
It's not only the preview .. if I insert a uploaded image with the attach tag, the image in not shown in the message body, it's shown as attach. In the message body I see the text **No message is associated with this attachment.**

![img01](https://cloud.githubusercontent.com/assets/2270223/14324094/ba095c7e-fc23-11e5-894c-d5a6907701c9.png)
![img02](https://cloud.githubusercontent.com/assets/2270223/14324149/f801bb98-fc23-11e5-892c-765a369ccf6b.png)

That all worked perfect in a older release !!!!

**Comment by MissAllSunday** [comment]
I committed a fix for it but theres another issue, currently  
showAttachment() function needs a topic ID, new attachments obviously  
doesn't have that so theres no choice but to split the query to get the  
attachment info into 2, one to get the atach info and another one to get  
the message info if the attachment has a msg associated with it.

Another alternative is to rely on the $_request['preview'] var and use  
another query that doesn't load the msg info  or use a nasty ternary  
inside our current query.

**Comment by PortaMx** [comment]
Well, your change worked as well .. but ..
There is a heavy bug is the **jquery.sceditor.smf.js** file, they I have fixed, so a inserted attach is correct shown.
Because I have currently no chance to make a pull request, I upload a zipped file here.
Also I fix the qoute time display for the SCE editor, to show as in the posted message (see attached image).

Topic the will qouted:
![quoted_topic](https://cloud.githubusercontent.com/assets/2270223/14354549/09d37d24-fcde-11e5-90c2-04691209596e.png)

The qouted topic in the editor:
![editor_fixes](https://cloud.githubusercontent.com/assets/2270223/14354444/9a33041c-fcdd-11e5-9fcc-9fd4a342ff57.png)

[jquery.sceditor.smf.zip](https://github.com/SimpleMachines/SMF2.1/files/208546/jquery.sceditor.smf.zip)

**Comment by PortaMx** [comment]
Here the simple fix for Preview a message ..
In the file **ShowAttachments.php** on line 88 change the query to:

```
        $request = $smcFunc['db_query']('', '
            SELECT a.id_folder, a.filename, a.file_hash, a.fileext, a.id_attach, a.attachment_type, a.mime_type, a.approved, m.id_member
            FROM {db_prefix}attachments AS a
                INNER JOIN {db_prefix}messages AS m ON (m.id_msg = a.id_msg AND m.id_topic = {int:current_topic} OR {int:current_topic} = 0)
                INNER JOIN {db_prefix}boards AS b ON (b.id_board = m.id_board AND {query_see_board})
            WHERE a.id_attach = {int:attach}
            LIMIT 1',
            array(
                'attach' => $attachId,
                'current_topic' => $attachTopic,
            )
        );
```

`
That worked ...

**Comment by PortaMx** [comment]
On more fix in **ShowAttachments.php** on line 133

replace
`if ($showThumb && $file['id_thumb'])
    {
        $request = $smcFunc['db_query']('', '
`
with
`if ($showThumb && !empty($file['id_thumb']))
    {
        $request = $smcFunc['db_query']('', '
`

**Comment by MissAllSunday** [comment]
Its not that simple, someone can easily bypass the query_see_board by  
setting $_REQUEST['topic'] to 0

   I solved it by splitting the query in two, once to get the attachment  
data and another one to get the message id and the check to  
query_see_board, the second query is only performed if the attachment has  
a message ID associated with it.

By splitting the queries we can have total control over which attachments  
need the query_see_board and which ones doesn't.

Theres a cache check so hopefully doing 2 queries isn't going to be that  
much of a problem.

**Comment by PortaMx** [comment]
Well .. Ok .. let my known, if you have make the fix, so I can test that ..

**Comment by MissAllSunday** [comment]
Heres the commit: https://github.com/MissAllSunday/SMF2.1/commit/3ec131e25aa40fdae5e562bd772da5ad4dd01a17

**Comment by PortaMx** [comment]
Well .. your changes are mostly good .. but a lot of smaller changes are need to get all correct to work.

Here comes the file with my changes .. all work perfect, also in the WYSIWYG editor ...

[ShowAttachments.zip](https://github.com/SimpleMachines/SMF2.1/files/209394/ShowAttachments.zip)

**Comment by MissAllSunday** [comment]
I agree on the space after the if but otherwise its not really necessary to cast all those ints, the DB layer does that for you. Perhaps one cast and check for is_int if we are being paranoid about it but thats it.

**Comment by MissAllSunday** [comment]
Another issue, we cannot rely on the filename to get a thumbnail, we  
already do a query and fetch the id_thumb row so its easier to just use  
that, its already there and its safer then relying on a filename and a  
hardcoded string

**Comment by PortaMx** [comment]
First .. the space after if is cosmetic ;-)
Second ..
I change the first query and add **thumb_id** .
In the second query (to get the thumb) i use the **thumb_id** from the first query to get the thumb image. Also the donwload counter is NOT updated is a thumb requested.

Here the modificated file
[ShowAttachments.zip](https://github.com/SimpleMachines/SMF2.1/files/211793/ShowAttachments.zip)

**Comment by MissAllSunday** [comment]
Already made a commit for this at https://github.com/MissAllSunday/SMF2.1/commit/ce7a675964df2f91d9c90e98592ee0e2af3d8312

**Comment by PortaMx** [comment]
Well .. I have updated the file again and optimize caching, so we need  less queries, because I cache the attach and also the thumnail. This will speed up the handling of attached images.
I have tested this on all situations and I can't find any problems ..

Update on 11. April, 17:40:
Now I put all images in one cachefile, this is faster.
here the optimized file:

[ShowAttachments.zip](https://github.com/SimpleMachines/SMF2.1/files/213330/ShowAttachments.zip)
